### PR TITLE
Guarantee ordering for SignalR client HttpConnection.send

### PR DESCRIPTION
**Pull request by pranavkm**
Resolves https://github.com/aspnet/AspNetCore/issues/9901

### CODE REVIEW

File: src/SignalR/clients/ts/signalr-protocol-msgpack/package.json [hunk 1]
```diff
@@ -42,7 +42,7 @@
     "src/**/*"
   ],
   "dependencies": {
-    "@aspnet/signalr": "file:../signalr",
+    "@aspnet/signalr": "link:../signalr",
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: Don't you need to update the msgpack npmproj now?

Comment (pranavkm) [reply src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: Blazor project?

Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr-protocol-msgpack/package.json]: https://github.com/aspnet/AspNetCore/blob/34a68d5f4860e1e5767839695e45f1eeda4c9f2d/src/SignalR/clients/ts/signalr-protocol-msgpack/signalr-protocol-msgpack.npmproj#L17

**Comment by dougbu** [comment]
This will need QB approval.

**Comment by BrennanConroy** [comment]
This is fixing a break in the official build so I think we can merge without approval

**Comment by natemcmaster** [comment]
@pranavkm can you explain the problem better for the rest of us who haven't looked at the build failure yet? I see official builds passing on all legs except the templates tests, and not sure I understand why.

Also, how did this error get merged in the first place? Is there a gap in PR validation?

> I think we can merge without approval

Let's get @Pilchie to make a call here.

**Comment by pranavkm** [comment]
@natemcmaster it passes the public build, but official builds seem to be broken: https://github.com/aspnet/AspNetCore-Internal/issues/2611. I'm not super sure what the difference is between the two, but I think this is a valid fix. 

There's a couple of issues here:

* `signalr-protocol-msgpack` used a `file` reference which was different from the `link` reference that the Functional test project uses. This caused duplicate references to SignalR to appear when restoring the functional test project

* The target file that edited and replaced the `package.json` file did not restore the original file. Consequently, it referenced a NPM package with the current build version that does not exist.

* Fixing these issues caused the `yarn.lock` file for the blazor project to change. We check in the code gen content for that project and have a CodeCheck.ps1 check that verifies that the files in that project are up to date. Changes to lock file caused it to fail the build.

**Comment by Pilchie** [comment]
In general, I'm always in favor of pre-emptively checking in a fix to un-break builds, as long as we then follow-up with an approval.

File: src/SignalR/clients/ts/signalr-protocol-msgpack/signalr-protocol-msgpack.npmproj [hunk 2]
```diff
@@ -14,7 +14,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <ReplacePackageContent Include="file:../signalr" ReplaceWith=">=$(PackageVersion)" />
+    <ReplacePackageContent Include="link:../signalr" ReplaceWith=">=$(PackageVersion)" />
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr-protocol-msgpack/signalr-protocol-msgpack.npmproj]: Change again :/

Comment (pranavkm) [reply src/SignalR/clients/ts/signalr-protocol-msgpack/signalr-protocol-msgpack.npmproj]: Back to being `links` everywhere.

File: src/Components/Browser.JS/utils/replace-paths.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const path = require("path");
```
Comment (pranavkm) [inline src/Components/Browser.JS/utils/replace-paths.js]: cc @BrennanConroy

File: src/Components/Browser.JS/package.json [hunk 4]
```diff
@@ -9,12 +9,13 @@
     "clean": "node utils/node_modules/rimraf/bin.js node_modules",
     "prebuild": "yarn run clean && yarn install --mutex network",
     "build": "yarn run build:debug && yarn run build:production",
-    "build:debug": "cd src && webpack --mode development --config ./webpack.config.js",
+    "build:debug": "cd src && webpack --mode development --config ./webpack.config.js & node ../utils/replace-paths.js",
```
Comment (BrennanConroy) [inline src/Components/Browser.JS/package.json]: Is `&&` required instead of `&`?

File: src/Components/Browser.JS/utils/replace-paths.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+const fs = require("fs");
+
+const serverJsPath = path.resolve(process.cwd(), "..", "dist", "Debug", "blazor.server.js");
+const signalRPath = path.resolve(process.cwd(), "..", "..", "..", "SignalR").replace(/\\/g, "/");
+const pathToReplace = "../../../SignalR";
+
+console.log(signalRPath);
```
Comment (BrennanConroy) [inline src/Components/Browser.JS/utils/replace-paths.js]: nit: remove

**Review by BrennanConroy — DISMISSED** [review]

**Comment by pranavkm** [comment]
No good. I'll merge the revert in. We'll try addressing this in preview7.

**Comment by pranavkm** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=211639&view=logs

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2623

**Review by BrennanConroy — APPROVED** [review]

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 6]
```diff
@@ -501,3 +511,96 @@ export class HttpConnection implements IConnection {
...
+    private sendBufferedData: PromiseSource;
+    private executing: boolean = true;
+    private transportResult: PromiseSource;
+
+    constructor(private readonly transport: ITransport) {
+        this.sendBufferedData = new PromiseSource();
+        this.transportResult = new PromiseSource();
+
+        // Suppress typescript error about handling Promises. The sendLoop doesn't need to be observed
+        this.sendLoop().then(() => {}, (_) => {});
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: It would be nice if this Promise was returned from TransportSendQueue.stop(), but I can clean that up in a later PR.

**Review by halter73 — APPROVED** [review]

**Review by dougbu — APPROVED** [review]
MSBuild changes look knarly but fine if they work 😺

**Comment by dougbu** [comment]
@mkArtakMSFT this is passing the CI checks now.

@pranavkm could you please double-check the produced packages to ensure the NPM things you're touching turn out :ok: at least in an unofficial build? Also recommend making an internal copy of your branch and queuing a kind-of official build using that branch. (Remove tags from that build to be sure no release pipeline kicks off. Tags are usually added very early in the build.)

**Comment by pranavkm** [comment]
> Also recommend making an internal copy of your branch and queuing a kind-of official build using that branch

How does one go about doing this?

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj [hunk 7]
```diff
@@ -7,23 +7,12 @@
...
-  <!-- Override the default 'Build' target from eng/targets/Npm.Common.targets. -->
-  <Target Name="Build"
-          Inputs="@(WebpackInputs)"
-          Outputs="@(WebPackOutputs)">
-    <RemoveDir Directories="dist" />
-
-    <Yarn Command="run build:debug" />
-    <Yarn Command="run build:production" />
-  </Target>
+  <Target Name="Restore" />
```
Comment (SteveSandersonMS) [inline src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: Could you add a comment about why this odd-looking target is here?

File: src/Components/Browser.JS/dist/.gitattributes [hunk 8]
```diff
@@ -0,0 +1,2 @@
+*.js    -diff -merge
+*.js    linguist-generated=true
```
Comment (SteveSandersonMS) [inline src/Components/Browser.JS/dist/.gitattributes]: Could you add comments to this file to say what these lines are for?

File: src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj [hunk 9]
```diff
@@ -50,10 +50,14 @@
...
-    <EmbeddedResource Include="..\..\Browser.JS\dist\$(Configuration)\blazor.server.js" LogicalName="_framework/%(Filename)%(Extension)" />
+  <PropertyGroup>
+    <BlazorServerJSFile>..\..\Browser.JS\dist\Release\blazor.server.js</BlazorServerJSFile>
+    <BlazorServerJSFile Condition="'$(Configuration)' == 'Debug' AND Exists('..\..\Browser.JS\dist\Debug\blazor.server.js')">..\..\Browser.JS\dist\Debug\blazor.server.js</BlazorServerJSFile>
+  </PropertyGroup>
 
-    <EmbeddedResource Include="..\..\Browser.JS\dist\$(Configuration)\blazor.server.js.map" LogicalName="_framework/%(Filename)%(Extension)" Condition="Exists('..\..\Browser.JS\dist\$(Configuration)\blazor.server.js.map')" />
+  <ItemGroup>
+    <EmbeddedResource Include="$(BlazorServerJSFile)" LogicalName="_framework/%(Filename)%(Extension)" />
+    <EmbeddedResource Include="$(BlazorServerJSFile).map" LogicalName="_framework/%(Filename)%(Extension)" Condition="Exists('$(BlazorServerJSFile).map')" />
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj]: I'd be interested to know why these changes were needed. No doubt it's for a good reason, but it's not clear to me on the surface why the previous logic was inadequate.

Comment (pranavkm) [reply src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj]: When I switched the package.json to use `link`, the Debug builds of blazor.*.js had absolute paths to SignalR embedded in them. The production builds did not since it does not include line mappings. I tried a couple of ways to solving it including rewriting the file after it was generated, but that didn't seem to help on the CI.

The change here is to say that production builds of the file are checked in and always available and we'd use Debug builds if they're available and we're building in Debug. What's missing is here a possible project reference to the npm proj which would help ensure building the server project in Debug would result in the debug files to be generated. But I wanted to sort out build incrementalism for the npmproj (preview7?) before we did that.

**Comment by pranavkm** [comment]
@SteveSandersonMS could I add the comment changes in the master branch?

**Comment by pranavkm** [comment]
@dougbu the build mostly succeeded. Failure is the same test failures we've seen elsewhere on Ubuntu: https://dnceng.visualstudio.com/internal/_build/results?buildId=212171

I verified the package contents. The most significant change I noted is that we now use Semver2.0 for the package versions:

```diff
-  "version": "3.0.0-preview5-19227-01",
+  "version": "3.0.0-preview6.19304.7",
```

@anurse it seems like it should be "OK", but just making sure you're aware of this.

**Comment by anurse** [comment]
:+1: yep, @natemcmaster gave me the heads-up this change was rolling out

**Comment by SteveSandersonMS** [comment]
> @SteveSandersonMS could I add the comment changes in the master branch?

Of course! I don't want to derail getting this merged, but just wanted the reasoning to be recorded because some of the changes are quite obscure.

**Comment by natemcmaster** [comment]
> > Also recommend making an internal copy of your branch and queuing a kind-of official build using that branch
> 
> How does one go about doing this?

Push your branch to https://dev.azure.com/dnceng/internal/_git/aspnet-AspNetCore and trigger a custom branch build on https://dev.azure.com/dnceng/internal/_build?definitionId=21

**Comment by pranavkm** [comment]
@natemcmaster already done: https://dnceng.visualstudio.com/internal/_build/results?buildId=212171
